lazy_tensor_core/lazy_tensor_core/core/optimize.py

```diff
@@ -0,0 +1,90 @@
+from lazy_tensor_core import _LAZYC
+import dataclasses
+from typing import List, Dict, Any, Callable
+import copy
+from torch import fx
+import torch
+
+debug = False
+
+@dataclasses.dataclass
+class GraphInputMatcher:
+    tensor_id_to_arg_idx: Dict[int, int]
+    graph_input_tensor_ids: List[int]
+    # there are 2 categories of graph_input_tensors.
+    # Category 1: those whose id are not found in tensor_id_to_arg_idx. These are
+    # most likely const tensors and we can get its content from graph_input_tensors
+    # Category 2: those whose id are found in tensor_id_to_arg_idx. We should get
+    #  the tensor from method arguments
+    graph_input_ivalues: List[Any]
+
+    # get the real graph input tensors
+    def __call__(self, args):
+        real_input = []
+        for tensor_id, traced_ivalue in zip(self.graph_input_tensor_ids, self.graph_input_ivalues):
+            arg_idx = self.tensor_id_to_arg_idx.get(tensor_id, None)
+            if arg_idx is None:
+                inp = traced_ivalue
+            else:
+                inp = args[arg_idx]
+            real_input.append(inp)
+        return real_input
+
+def force_lazy_device(model: fx.GraphModule):
+    """
+    Factory methods in a Fx graph may create tensors for a specific eager devices.
+    If we take no actions, those eager tensors will be mixed with lazy tensors and
+    cause crash. This method overwrite those eager device to lazy device.
+    """
+    def tolazydevice(dev):
+        if isinstance(dev, torch.device):
+            return torch.device("lazy", index=dev.index)
+        return dev
+
+    for nd in model.graph.nodes:
+        nd.args = tuple(tolazydevice(arg) for arg in nd.args)
+        nd.kwargs = {k: tolazydevice(v) for k, v in nd.kwargs.items()}
+    model.recompile()
+
+
+def optimize(model: fx.GraphModule, example_inputs) -> Callable:
```

Comment:
perhaps 'optimize' is too commonly used and not clear what it means.

this function essentially performs lazy tracing over a FX graph right?
-> maybe name lazy_trace_fx()?

the file name is also similar; if this code is all specific to lazy-fx, maybe name the file `lazy_fx.py`?

Reply:
I totally agree that the name optimize is too general. Honestly I had a hard time to pick the method name when I wrote the PR.

The main point of this method is to reuse the compiled graph without retracing. Calling it 'lazy_trace_fx' seem not calling that out.  How about  `extract_compiled_graph`  or `extract_compiled_graph_from_fx` ? I slightly prefer the former than the later since it's shorter and it's not necessary to emphasis the Fx aspect.

What do you think?